### PR TITLE
fix(discord): prevent uncaught exception on stale-socket reconnect

### PR DESCRIPTION
**Pull request by Lidang-Jiang**
## Summary

- Problem: Discord gateway lifecycle's `onAbort()` sets `reconnect.maxAttempts = 0` then calls `disconnect()`, which triggers Carbon's internal close handler. The close handler tries to reconnect, hits the 0-attempt limit, and emits an error with no listener — crashing the process
- Why it matters: This is an uncaught exception that crashes the entire OpenClaw process, affecting all connected channels
- What changed: Strip WebSocket close/error listeners before disconnecting in `onAbort()` to prevent the stale-socket close event from triggering the error path. A no-op error listener is retained to prevent Node from throwing if the socket emits "error" during disconnect. Guard: only strip when no reconnect is already in flight
- What did NOT change (scope boundary): Normal reconnect flow unaffected, `disconnectGatewaySocketWithoutAutoReconnect`'s drain promise still resolves correctly

Fixes #55618

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [x] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #55618
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: `onAbort()` calls `disconnect()` after setting `maxAttempts = 0`, but `disconnect()` closes the WebSocket which fires Carbon's `handleClose`. `handleClose` attempts reconnect, hits the 0 limit, and emits an unhandled error event
- Missing detection / guardrail: No test verified the abort → disconnect → stale-socket-close path
- Prior context: Carbon's WebSocket handling assumes listeners are always present when close events fire
- Why this regressed now: Likely a latent issue surfaced by network instability or gateway restart timing

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
- Target test or file: `extensions/discord/src/monitor/provider.lifecycle.test.ts`
- Scenario the test should lock in: After `onAbort()`, Carbon's close listeners must be removed, a no-op error listener retained, and late `socket.emit("close")` must not throw
- Why this is the smallest reliable guardrail: The bug is in listener management — checking listenerCount after abort is the most direct assertion
- Existing test that already covers this (if any): None

## User-visible / Behavior Changes

- Discord stale-socket close events during lifecycle teardown no longer crash the process
- No change to normal reconnect behavior

## Diagram (if applicable)

```text
Before:
[onAbort] -> maxAttempts=0 -> disconnect() -> socket.close fires
  -> Carbon handleClose -> reconnect attempt -> maxAttempts exceeded
  -> error emitted -> NO LISTENER -> uncaught exception -> PROCESS CRASH

After:
[onAbort] -> strip close/error listeners -> add no-op error listener
  -> maxAttempts=0 -> disconnect() -> socket.close fires
  -> no close listeners -> event ignored -> safe teardown
  (if socket emits "error" during disconnect -> no-op listener absorbs it)
```

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No

### Environment

- OS: Linux
- Runtime/container: Node 22+
- Integration/channel (if any): Discord

### Steps

1. Connect OpenClaw to Discord
2. Trigger gateway lifecycle abort (e.g., network interruption during reconnect)
3. Stale WebSocket emits close event after lifecycle teardown

### Expected

- Process continues running, stale socket event ignored

### Actual

- Before fix: `Max reconnect attempts (0)` uncaught exception, process crashes
- After fix: Stale socket event harmlessly ignored

## Evidence

- [x] Failing test/log before + passing after

<details>
<summary>Before (without fix — new test fails)</summary>

```
> openclaw@2026.3.26 test:extension /ssd1/jianglidang/workspace/openclaw/worktrees/fix-discord-crash
> node scripts/test-extension.mjs discord

[test-extension] Running 94 test files for discord with vitest.channels.config.ts
[test-extension] Split into 43 shared and 51 isolated files

 RUN  v4.1.2 /ssd1/jianglidang/workspace/openclaw/worktrees/fix-discord-crash

 Test Files  41 passed (41)
      Tests  325 passed (325)
   Start at  17:57:16
   Duration  10.60s (transform 16.33s, setup 56.89s, import 210ms, tests 4.22s, environment 4ms)


 RUN  v4.1.2 /ssd1/jianglidang/workspace/openclaw/worktrees/fix-discord-crash

 ❯ extensions/discord/src/monitor/provider.lifecycle.test.ts (22 tests | 1 failed) 6701ms
     × strips socket close/error listeners on abort so stale-socket close does not crash 297ms

⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯

 FAIL  extensions/discord/src/monitor/provider.lifecycle.test.ts > runDiscordGatewayLifecycle > strips socket close/error listeners on abort so stale-socket close does not crash
AssertionError: expected 1 to be +0 // Object.is equality

- Expected
+ Received

- 0
+ 1

 ❯ extensions/discord/src/monitor/provider.lifecycle.test.ts:885:43
    883|     // After abort, the WebSocket's close/error listeners should have …
    884|     // removed to prevent Carbon's handleClose from firing during tear…
    885|     expect(socket.listenerCount("close")).toBe(0);
       |                                           ^
    886|     expect(socket.listenerCount("error")).toBe(0);
    887|

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯


 Test Files  1 failed | 50 passed (51)
      Tests  1 failed | 486 passed (487)
   Start at  17:57:27
   Duration  41.91s (transform 17.93s, setup 189.80s, import 2.45s, tests 40.93s, environment 6ms)

 ELIFECYCLE  Command failed with exit code 1.
```

</details>

<details>
<summary>After (with fix — all 22 lifecycle tests pass)</summary>

```
> openclaw@2026.3.26 test /ssd1/jianglidang/workspace/openclaw/worktrees/fix-discord-crash
> node scripts/test-parallel.mjs -- extensions/discord/src/monitor/provider.lifecycle.test.ts

[test-parallel] start channels-provider.lifecycle-isolated workers=2 filters=1

 RUN  v4.1.2 /ssd1/jianglidang/workspace/openclaw/worktrees/fix-discord-crash


 Test Files  1 passed (1)
      Tests  22 passed (22)
   Start at  18:41:59
   Duration  11.89s (transform 1.61s, setup 4.73s, import 68ms, tests 6.93s, environment 0ms)

[test-parallel] done channels-provider.lifecycle-isolated code=0 elapsed=12.9s
```

</details>

## Human Verification (required)

- Verified scenarios: New test verifies listenerCount drops to 0 for close (Carbon's listeners removed) and 1 for error (no-op safety listener retained); late `socket.emit("close")` does not throw
- Edge cases checked: Abort with reconnect in flight (listeners preserved), abort without reconnect (listeners stripped + no-op error listener added)
- What you did **not** verify: Live Discord bot with real network interruption (test-only verification)

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

### Addressed Review Feedback

- **codex P1 (missing error listener)**: After stripping all error listeners, a no-op `() => {}` error listener is now added to prevent Node `EventEmitter` from throwing an uncaught exception if the WebSocket emits "error" during disconnect. This matches the `ignoreSocketError` pattern used in `disconnectGatewaySocketWithoutAutoReconnect`.

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? No
- Migration needed? No

## Risks and Mitigations

- Risk: Stripping listeners could prevent `disconnectGatewaySocketWithoutAutoReconnect`'s drain promise from resolving if a reconnect is in flight
  - Mitigation: Guard condition `!reconnectInFlight` ensures listeners are only stripped when safe

🤖 AI-assisted (Claude Opus 4.6), fully tested

### CODE REVIEW

File: extensions/discord/src/monitor/provider.lifecycle.reconnect.ts [hunk 1]
```diff
@@ -402,6 +402,27 @@ export function createDiscordGatewayReconnectController(params: {
...
+    // close event, its own listener must stay so the drain promise resolves.
+    if (!reconnectInFlight) {
+      const socket = params.gateway.ws;
+      if (socket) {
+        for (const listener of socket.listeners("close")) {
+          socket.removeListener("close", listener);
+        }
+        for (const listener of socket.listeners("error")) {
+          socket.removeListener("error", listener);
+        }
```
[the remark this answers is not in the document]
Comment (Lidang-Jiang) [reply extensions/discord/src/monitor/provider.lifecycle.reconnect.ts]: Already handled. The `onAbort` handler (lines 414–430 of `provider.lifecycle.reconnect.ts`) follows the same safe sequence as the reconnect path:

1. Guards with `if (!reconnectInFlight)` to avoid interfering with an active reconnect that has its own listeners
2. Strips all `close` and `error` listeners from the socket
3. Immediately re-registers a no-op error listener: `socket.on("error", () => {})` (line 425)
4. Only then calls `params.gateway.disconnect()`

This guarantees there is always an `error` listener present when `disconnect()` is called, preventing uncaught exceptions. The comment at line 420 explains the rationale.

**Comment by steipete** [comment]
Closing this as implemented after Codex review.

Current `main` already handles the stale Discord gateway teardown/reconnect crash through the newer gateway supervisor path, so this PR is redundant.

What I checked:

- **Abort path keeps handlers through disconnect:** `waitForDiscordGatewayStop()` calls `gateway.disconnect()` before lifecycle cleanup, with an explicit comment that late `error` events during disconnect must stay handled instead of becoming uncaught. ([`extensions/discord/src/monitor.gateway.ts:18`](https://github.com/openclaw/openclaw/blob/d3d9b5738f83/extensions/discord/src/monitor.gateway.ts#L18), [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83))
- **Late teardown errors are suppressed:** `createDiscordGatewaySupervisor()` classifies `Max reconnect attempts...` as `reconnect-exhausted` and suppresses those events in `teardown`/`disposed` phases instead of letting them escape. ([`extensions/discord/src/monitor/gateway-supervisor.ts:140`](https://github.com/openclaw/openclaw/blob/d3d9b5738f83/extensions/discord/src/monitor/gateway-supervisor.ts#L140), [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83))
- **Regression test covers teardown case:** `gateway-supervisor.test.ts` asserts that after `detachLifecycle()`, emitting `Max reconnect attempts (0)...` does not throw and is logged as a suppressed teardown error. ([`extensions/discord/src/monitor/gateway-supervisor.test.ts:59`](https://github.com/openclaw/openclaw/blob/d3d9b5738f83/extensions/discord/src/monitor/gateway-supervisor.test.ts#L59), [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83))
- **Regression test covers dispose/startup-cleanup case:** `provider.test.ts` asserts startup cleanup disconnects the shared gateway, then a late `Max reconnect attempts (0)...` emitter error is suppressed after dispose rather than crashing. ([`extensions/discord/src/monitor/provider.test.ts:281`](https://github.com/openclaw/openclaw/blob/d3d9b5738f83/extensions/discord/src/monitor/provider.test.ts#L281), [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83))
- **Lifecycle uses the new stop/supervisor flow:** `runDiscordGatewayLifecycle()` now routes shutdown through `waitForDiscordGatewayStop()` with the shared `gatewaySupervisor`, so the old PR's direct socket-listener surgery is no longer the active seam. ([`extensions/discord/src/monitor/provider.lifecycle.ts:441`](https://github.com/openclaw/openclaw/blob/d3d9b5738f83/extensions/discord/src/monitor/provider.lifecycle.ts#L441), [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83))

So I’m closing this as already implemented rather than keeping a duplicate issue open.

Review notes: reviewed against [d3d9b5738f83](https://github.com/openclaw/openclaw/commit/d3d9b5738f83d92787f57db9186ddcebda60af9a).
